File: 0242-valid-anagram/main.cpp
// C++ hash hable holution
class Solution {
public:
    bool isAnagram(string s, string t) {
        unordered_map<char,int> hashmap;
        if (s.length() != t.length()) {
            return false;
        }
        for (int i = 0; i < s.length(); i++) {
            if (hashmap.find(s[i]) != hashmap.end()) { // it exists
                hashmap[s[i]]++;
            } else {
                hashmap[s[i]] = 1;
            }
            if (hashmap.find(t[i]) != hashmap.end()) { // it exists
                hashmap[t[i]]--;
            } else {
                hashmap[t[i]] = -1;
            }
        }
        // using iterators to check each item
        for (auto it = hashmap.begin(); it != hashmap.end(); it++) {
            if (it->second != 0) {
                return false;
            }
        }
        return true;
    }
};

File: 0242-valid-anagram/main.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # hashmap
        # time: O(N)
        # space: O(N)
        hashmap = {}
        for i in s:
            if i not in hashmap:
                hashmap[i] = 0
            hashmap[i] += 1
        for j in t:
            if j not in hashmap:
                hashmap[j] = 0
            hashmap[j] -= 1
        for k,v in hashmap.items():
            if v != 0:
                return False
        return True
